### PR TITLE
Sofastream: read specific data from sofa-file

### LINKED ISSUE #58: Add support to open/enter a sofa file without reading the data
For large files reading the entire file into memory is not always economic.
To solve this, we could support opening sofa files without reading the data into memory using the __enter__ method.
This is already supported by netcdf
``` python
with Dataset(filename, "r", format="NETCDF4") as file:
    # read the data here
```
In addition we could use support netcdfs slicing support for arrays to improve this even further, i.e. for channel wise reading of data or reading only partial spherical harmonic orders, etc.

### PR DESCRIPTION
**Pull request by hoyer-a**
sofar class to read specific data from sofa-file without reading entire file into memory.

closes #58 

- We should avoid having large SOFA files in the repository. Whats the easiest way of getting it out of the history?
    - use `tempfile` to create a (not large) sofa file that is writte jsut for testing and not saved. `prune` history to avoid pushing the large file to develop
- Do we want to manipulate the string representation of variables that are returned if not slicing
    - we decided not to do this, because a netcdf variable is returned in any case. The slicing happens outside of SofaStream. 
- Do we want to add a Sofastream.inspect() function in this PR (I think its easy to implement and can partly reuse code that you already wrote)
    - Yes, we want that. The output should ideally be identical to `Sofa.inspect()`
- Should we change the name to SofaStream?
    - yes, we want

### CODE REVIEW

**Comment by hoyer-a** [comment]
Any idea what's the matter with the failing tests and the conventions? Is the conventions.rst being updated automatically?

**Comment by ahms5** [comment]
The test compares the conventions on github and on sofaconventions.org. on main it also fails, parsing data from websites is always critical, sometimes the website changes slidely and noting is working anymore. The problem seems to be the parsing from github, cloning the repo, would be a more stable solution, but it would also use more resources. it would be an other PR

**Comment by hoyer-a** [comment]
Thanks for the explanation!

**Comment by f-brinkmann** [comment]
That right. I noted it a little while ago but did not find time to change it. Its already the second time that this happens and I think swithcing to cloning will make sense here.

**Comment by f-brinkmann** [comment]
Thanks, works like a charm. There are a few things we could discuss at the weekly before doing detailed reviews:

- We should avoid having large SOFA files in the repository. Whats the easiest way of getting it out of the history?
- Do we want to manipulate the string representation of variables that are returned if not slicing
- Do we want to add a Sofastream.inspect() function in this PR (I think its easy to implement and can partly reuse code that you already wrote)
- Should we change the name to SofaStream?

**Comment by hoyer-a** [comment]
Just a quick response concerning the string representation. Since the returned variables are netCDF-variables and no Sofastream objects the netCDF4._netCDF4.Variable.__repr__() is called when printing, which is read only. 
But maybe there's another way, would definitely be nice!

File: sofar/sofastream.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        self._file.close()
+
+    def __getattr__(self, name):
+        # get netCDF4-attributes and -variable-keys from sofa-file
+        dset_variables = np.array([key for key in self._file.variables.keys()])
+        dset_attributes = np. asarray(self._file.ncattrs())
+
+        # remove delimiter from passed sofar-attribute
+        name = name.replace('_', '')
+        name = name.replace('GLOBAL', '')
```
[the remark this answers is not in the document]
Comment (hoyer-a) [reply sofar/sofastream.py:93]: Thanks! Way more elegant this way

File: sofar/sofastream.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        dset_attributes = np.asarray(self._file.ncattrs())
+
+        # remove delimiter from passed sofar-attribute
+        name_netcdf = name.replace(
+            'GLOBAL_', '').replace('Data_', 'Data.')
+
+        # Handle variable-attributes (e.g. '_Units' and '_Type')
+        var_attr = None
+        if "_Units" in name_netcdf or "_Type" in name_netcdf:
+            name_split = name_netcdf.split('_')
```
Comment (f-brinkmann) [inline sofar/sofastream.py:98]: This can be shorter if it always produces a list with two entries
```suggestion
            name_netcdf, var_attr = name_netcdf.split('_')
```
If there might be more entries it could be
```suggestion
            name_netcdf, var_attr = name_netcdf.split('_')[0:2]
```

File: sofar/sofastream.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+            if var_attr is not None:
+                self._data = getattr(self._data, var_attr)
+
+        # get value if passed attribute points to a netCDF4-attribute
+        elif name_netcdf in dset_attributes:
+            # get attribute value from SOFA-file
+            self._data = self._file.getncattr(name_netcdf)
+
+        else:
+            raise ValueError("attribute is not in dataset")
```
Comment (f-brinkmann) [inline sofar/sofastream.py:115]: Maybe be more specific
```suggestion
            raise AttributeError(f"{name} is not contained in SOFA file")
```

File: sofar/sofastream.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    - If a variable is called, a netCDF4-variable is returned. To get the
+      values it needs to be sliced (see examples).
+
+    Examples
+    --------
+    Get an attribute from a SOFA-file:
+
+        >>> from sofar import SofaStream
+        >>> filename = "path/to/file.sofa"
+        >>> with SofaStream(filename) as file:
```
Comment (f-brinkmann) [inline sofar/sofastream.py:45]:
```suggestion
        >>> import sofar as sf
        >>> filename = "path/to/file.sofa"
        >>> with sf.SofaStream(filename) as file:
```

File: sofar/sofastream.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+            raise ValueError("attribute is not in dataset")
+
+        return self._data
+
+    def inspect(self, file=None):
+        """
+        Get information about the data inside a SOFA-file
+
+        Prints all information about attributes, variables and their
+        shape, dimensions and values that are contained in a SOFA-file.
```
Comment (f-brinkmann) [inline sofar/sofastream.py:124]: Maybe use the same docstring as in Sofa.inspect()
```suggestion
        Prints the values of all attributes and variables with six or less
        entries and the shapes and type of all numeric and string variables.
        When printing the values of arrays, single dimensions are discarded for
        easy of display, i.e., an array of shape (1, 3, 2) will be displayed as
        an array of shape (3, 2).
```

File: sofar/sofastream.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        dset_variables = np.array([key for key in self._file.variables.keys()])
+        dset_attributes = np.asarray(self._file.ncattrs())
+
+        # remove delimiter from passed sofar-attribute
+        name_netcdf = name.replace(
+            'GLOBAL_', '').replace('Data_', 'Data.')
+
+        # Handle variable-attributes (e.g. '_Units' and '_Type')
+        var_attr = None
+        if "_Units" in name_netcdf or "_Type" in name_netcdf:
```
Comment (f-brinkmann) [inline sofar/sofastream.py:97]: I think after converting to a netcdf name, any '_' will indicate you are looking at a variable attribute. Does the following work? It would be more general;
```suggestion
        if "_" in name_netcdf:
```

File: tests/test_sofastream.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+@pytest.fixture
+def temp_sofa_file(tmp_path_factory):
+
+    filename = tmp_path_factory.mktemp("data") / "test_sofastream.sofa"
+    sofa = sf.Sofa("SimpleFreeFieldHRIR")
+    sofa.Data_IR = np.array([[0, 1], [2, 3], [4, 5]])
+    sofa.GLOBAL_RoomType = "free field"
+    sofa.Data_SamplingRate_Units = "hertz"
+    sf.write_sofa(filename, sofa)
+    return filename
```
Comment (f-brinkmann) [inline tests/test_sofastream.py:20]: The fixture should go into conftest.py in the same directory. You can check it out in the pyfar package

File: sofar/sofastream.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
+from netCDF4 import Dataset
+import numpy as np
+
+class SofaStream():
+    """
+    Get specific data from SOFA-file without reading entire file into memory.
```
Comment (f-brinkmann) [inline sofar/sofastream.py:6]:
```suggestion
    Get data from SOFA-file without reading entire file into memory.
```

File: sofar/sofastream.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+
+    :class:`SofaStream` opens a SOFA-file and retreives only the requested
```
Comment (f-brinkmann) [inline sofar/sofastream.py:8]:
```suggestion
    :class:`SofaStream` opens a SOFA-file and retrieves only the requested
```

File: sofar/sofastream.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    data. To access certain data just append its name to your
+    :class:`SofaStream`-instance, e.g. ``'.GLOBAL_RoomType'``.
+    :class:`SofaStream` uses the namespace of sofar – for more information on
+    available data and its associated names refer to the :ref:`conventions
+    section <conventions_introduction>`. Also make sure to check the examples
+    below on how to use :class:`SofaStream`.
```
Comment (f-brinkmann) [inline sofar/sofastream.py:14]: I think I would be more brief here. The examples are doing a good job of explaining the class
```suggestion
    data. See the examples
    below on how to use :class:`SofaStream`.
```

File: sofar/sofastream.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+    SofaStream is supposed to be used within a with-statement. To acces data
+    append the corresponding name to your instance of :class:`SofaStream`.
+
+    Depending on the type of the requested data you either get the values
+    directly or need to slice the returned variable:
+
+    - If an attribute is called, its value is returned directly.
+    - If a variable is called, a netCDF4-variable is returned. To get the
+      values it needs to be sliced (see examples).
```
Comment (f-brinkmann) [inline sofar/sofastream.py:37]: I would suggest to remove this completely in favor of the examples.

File: tests/test_sofastream.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        # variable values
+        isinstance(var_data, np.ma.core.MaskedArray)
+        np.testing.assert_array_equal(var_data.squeeze(),
+                                      np.array([[0, 1], [2, 3], [4, 5]]))
+        assert var_attr == 'hertz'
+        # attribute values
+        isinstance(att_data, str)
+        assert att_data == "free field"
+
+def test_sofastream_warnings(temp_sofa_file):
```
Comment (f-brinkmann) [inline tests/test_sofastream.py:44]:
```suggestion
def test_sofastream_attribute_error(temp_sofa_file):
```

**Review by f-brinkmann — CHANGES_REQUESTED** [review]
This is really nice already. Most of my comments are structural and code stlye.

I noticed there are a couple of flake8 violations. You can for that by running `flake8 sofar tests` and you immediatly see them in your IDE if you install the Flake8 plugin.

File: docs/resources/conventions.rst [hunk 13]
```diff
@@ -88,164 +88,164 @@ Current
      - Comment
    * - GLOBAL_Conventions (*attribute*)
      - SOFA
-     - 
+     -
```
Comment (ahms5) [inline docs/resources/conventions.rst:91]: this file should be deleted on git since its always recreated if the doc os build. We sould then also add this file to the gitignore.

File: sofar/sofastream.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+    below on how to use :class:`SofaStream`.
+
+    Parameters
+    ----------
+    filename : str
+        Full path to a SOFA-file
+
+    Returns
+    --------
+    sofastream : SofaStream
```
Comment (ahms5) [inline sofar/sofastream.py:23]:
```suggestion
    sofa_stream : SofaStream
```
usually the naming convention is with underscores.

File: sofar/sofastream.py [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+
+    Notes
+    -----
+    Accessing data:
+
+    SofaStream is supposed to be used within a with-statement. To acces data
```
Comment (ahms5) [inline sofar/sofastream.py:29]:
```suggestion
    SofaStream is supposed to be used within a with-statement. To access data
```

File: sofar/sofastream.py [hunk 16]
```diff
@@ -0,0 +1,192 @@
+from netCDF4 import Dataset
+import numpy as np
+
+class SofaStream():
```
Comment (ahms5) [inline sofar/sofastream.py:4]:
```suggestion
class SofaStream(Sofa):
```
which doesnt it enherit from the sofa class, then it could share all functionalities. If I remember correctly, it should behave like a real Sofa class exept its reading from file and not from ram, and also the data is a netBla object if you dont slice. Then all the tests with sofa should also work with SofaStream.

File: sofar/sofastream.py [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+        >>>     data = file.Data_IR
+        >>>     print(data)
+        <class 'netCDF4._netCDF4.Variable'>
+        float64 Data.IR(M, R, N)
+        unlimited dimensions:
+        current shape = (11950, 2, 256)
+        filling on, default _FillValue of 9.969209968386869e+36 used
+
+    What is returned is a `netCDF-variable`. To access the values (in this
+    example the IRs) the variable needs to be sliced:
```
Comment (ahms5) [inline sofar/sofastream.py:62]: we can add here, that if you want to use all the data you should go for the Sofa class instead

File: sofar/sofastream.py (hunk 5 above)
Comment (ahms5) [reply sofar/sofastream.py:124]: I don't understand why we simplify the shape form (1,2,3) to (2,3). I think its not related to this pr. Is it just to display or is it also simplified on the data itself

File: tests/test_sofastream.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+from pytest import raises
+import pytest
+import netCDF4
+import numpy as np
+import os
+
+# Temporary SOFA-file
+@pytest.fixture
+def temp_sofa_file(tmp_path_factory):
+
```
Comment (ahms5) [inline tests/test_sofastream.py:13]: please document the fixture

File: tests/test_sofastream.py [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+        isinstance(att_data, str)
+        assert att_data == "free field"
+
+def test_sofastream_warnings(temp_sofa_file):
+
+    with SofaStream(temp_sofa_file) as file:
+        with raises(ValueError, match="attribute is not in dataset"):
+            file.Wrong_Attribute
+
+def test_sofastream_inspect(capfd, temp_sofa_file):
```
Comment (ahms5) [inline tests/test_sofastream.py:50]: maybe we can compare it with the result from Sofa class

**Review by ahms5 — CHANGES_REQUESTED** [review]
Thank you for your effort. looks already very good. I'm just wondering about the inheritance from the Sofa class.

File: sofar/sofastream.py (hunk 16 above)
Comment (hoyer-a) [reply sofar/sofastream.py:5]: Not sure about this, since SofaStream is just imitating the attribute/variable-access and `inspect()` of the `Sofa` class.

Comment (f-brinkmann) [reply sofar/sofastream.py:5]: No I think you're right. Inheriting does't work here. The `Sofa` class holds all the data of a sofa-file while `SofaStream` reads it upon request. Due to this even the `inspect` function that is contained in both classes has partly different code.

It might be possible to do this but it would require quite a large code refactoring.

File: sofar/sofastream.py [hunk 20]
```diff
@@ -71,6 +53,8 @@ class SofaStream():
         >>>     specific_irs = data[:,0,:]
         >>>     print(specific_irs.shape)
         (11950, 256)
+
+    If you want to use all the data from a SOFA-file use :class:`SOFA` instead.
```
Comment (f-brinkmann) [inline sofar/sofastream.py:57]: Maybe move this hint up right before the Parameters section and add a corresponding hint to the `Sofa` class?

**Review by f-brinkmann — CHANGES_REQUESTED** [review]
Looks great. I only had one small comment left for the docstring. Thanks a lot!

File: sofar/sofastream.py (hunk 5 above)
Comment (f-brinkmann) [reply sofar/sofastream.py]: I think that is bad decision that I made when I was young and wanted things to be as in Matlab

File: sofar/update_conventions.py
Comment (f-brinkmann) [inline sofar/update_conventions.py:1]: Not sure why these changes show up here. They are from #85, which was merged into develop and main

**Comment by hoyer-a** [comment]
I added a short introduction into `Sofa` and `SofaStream` to `sofar.Sofa.rst`. But for some reason the changes aren't included in the latest docs build...

**Comment by f-brinkmann** [comment]
It seems your update was somwhow not pushed - its working now :)

**Review by f-brinkmann — APPROVED** [review]
Looks good to me!

File: sofar/sofastream.py [hunk 21]
```diff
@@ -0,0 +1,178 @@
+from netCDF4 import Dataset
+import numpy as np
+
+
+class SofaStream():
+    """
+    Get data from SOFA-file without reading entire file into memory.
```
Comment (ahms5) [inline sofar/sofastream.py:7]:
```suggestion
    Read desired data from SOFA-file directly from disk without loading entire file into memory.
```

File: sofar/sofastream.py [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+
+    :class:`SofaStream` opens a SOFA-file and retrieves only the requested
+    data. See the examples below on how to use :class:`SofaStream`.
```
Comment (ahms5) [inline sofar/sofastream.py:10]:
```suggestion
    data.
```

File: sofar/sofastream.py [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+    class and :func:`read_sofa` function instead.
+
+    Parameters
+    ----------
+    filename : str
+        Full path to a SOFA-file
+
+    Returns
+    --------
+    sofa_stream : SofaStream
```
Comment (ahms5) [inline sofar/sofastream.py:22]:
```suggestion
    sofa_stream : SofaStream
        A SofaStream object which read directly from the file.
```

**Review by ahms5 — CHANGES_REQUESTED** [review]
Thanks, just a small comments on the doc

**Comment by hoyer-a** [comment]
Before merging i need to cleanup the history again.

File: sofar/sofastream.py (hunk 16 above)
Comment (f-brinkmann) [reply sofar/sofastream.py:5]: From my side, the current implementation is fine considering that we now have a general introduction for the module that explains Sofa and SofaStream. In addition users would get an AttrubiteError if for example trying to call `add_attribute`.

**Review by ahms5 — APPROVED** [review]
